From 67045f902666ca3f720f07b5255cdcc119b37c43 Mon Sep 17 00:00:00 2001
From: Khai Wen Ng <khai.wen.ng@intel.com>
Date: Fri, 12 Sep 2025 04:58:59 +0800
Subject: [PATCH 59/61] media: pci: intel: ipu6-isys-csi2: get link frequency
 from pad

In case driver does not have v4l2-control that stores link frequency,
try to retrieve link frequency from pad

Signed-off-by: Khai Wen Ng <khai.wen.ng@intel.com>
---
 drivers/media/pci/intel/ipu6/ipu6-isys-csi2.c | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/drivers/media/pci/intel/ipu6/ipu6-isys-csi2.c b/drivers/media/pci/intel/ipu6/ipu6-isys-csi2.c
index 6030bd23b4b94..0f5d2557db32b 100644
--- a/drivers/media/pci/intel/ipu6/ipu6-isys-csi2.c
+++ b/drivers/media/pci/intel/ipu6/ipu6-isys-csi2.c
@@ -80,6 +80,9 @@ static const struct ipu6_csi2_error dphy_rx_errors[] = {
 s64 ipu6_isys_csi2_get_link_freq(struct ipu6_isys_csi2 *csi2)
 {
 	struct media_pad *src_pad;
+	struct v4l2_subdev *ext_sd;
+	struct device *dev;
+	s64 ret;
 
 	if (!csi2)
 		return -EINVAL;
@@ -92,7 +95,20 @@ s64 ipu6_isys_csi2_get_link_freq(struct ipu6_isys_csi2 *csi2)
 		return PTR_ERR(src_pad);
 	}
 
-	return v4l2_get_link_freq(src_pad, 0, 0);
+	ext_sd = media_entity_to_v4l2_subdev(src_pad->entity);
+	if (WARN(!ext_sd, "Failed to get subdev for %s\n", csi2->asd.sd.name))
+		return -ENODEV;
+
+	ret = v4l2_get_link_freq(ext_sd->ctrl_handler, 0, 0);
+	if (ret) {
+		ret = v4l2_get_link_freq(src_pad, 0, 0);
+		if (ret) {
+			dev_err(dev, "Failed to get link frequency for %s\n",
+				csi2->asd.sd.name);
+			return ret;
+		}
+	}
+	return ret;
 }
 
 static int csi2_subscribe_event(struct v4l2_subdev *sd, struct v4l2_fh *fh,
-- 
2.50.1

